paddle/fluid/operators/detail/brpc_server.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/operators/detail/brpc_server.h"
+#include "paddle/fluid/operators/detail/request_handler.h"
+
+namespace sendrecv {
+class BRPCServiceImpl : public SendRecvService {
+ public:
+  BRPCServiceImpl(const std::unordered_map<
+                  std::string, paddle::operators::detail::RequestHandler*>&
+                      rpc_call_map) {
+    request_send_h_ = nullptr;
+    request_get_h_ = nullptr;
+    request_prefetch_h_ = nullptr;
+
+    auto it = rpc_call_map.find(paddle::operators::detail::kRequestSend);
+    if (it != rpc_call_map.end()) {
+      request_send_h_ = it->second;
+    }
+
+    it = rpc_call_map.find(paddle::operators::detail::kRequestSend);
+    if (it != rpc_call_map.end()) {
+      request_get_h_ = it->second;
+    }
+
+    it = rpc_call_map.find(paddle::operators::detail::kRequestPrefetch);
+    if (it != rpc_call_map.end()) {
+      request_prefetch_h_ = it->second;
+    }
+  }
+
+  virtual ~BRPCServiceImpl() {}
+
+  void SendVariable(google::protobuf::RpcController* cntl_butil,
+                    const VariableMessage* request, VoidMessage* response,
+                    google::protobuf::Closure* done) override {
+    brpc::ClosureGuard done_guard(done);
+
+    paddle::framework::Scope* local_scope = request_send_h_->scope();
```

Comment:
`request_send_h_` may be `nullptr`? well in some rare cases, but `request_prefetch_h_` may be null in most cases. Better to add a check here?

Reply:
Set `nullptr` when initialize.

Reply:
Well, if `request_prefetch_h_` is not set actually, `request_prefetch_h_->scope()` will cause error.

Reply:
I add checker when they're called.